File: Interview_Kit/Search/Pairs.java
package Search;

import java.io.*;
import java.util.*;

public class Pairs {

    // Complete the pairs function below.
    static int pairs(int k, int[] arr) {

        Arrays.sort(arr);
        int i = 0;
        int j = 1;
        int answer = 0;
        while(true){
            if(arr[j] - arr[i] > k) {
                if(j - i == 1) j++;
                if(j == arr.length) break; 
                i++;
            }else if(arr[j] - arr[i] == k){
                i++;
                if(j != arr.length - 1) j++;
                answer++;
                if(j == i) break;
            }else{
                if(j != arr.length - 1) j++;
                else break;
            }
        }
        return answer;
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        //BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        String[] nk = scanner.nextLine().split(" ");

        int n = Integer.parseInt(nk[0]);

        int k = Integer.parseInt(nk[1]);

        int[] arr = new int[n];

        String[] arrItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            int arrItem = Integer.parseInt(arrItems[i]);
            arr[i] = arrItem;
        }

        int result = pairs(k, arr);

        //bufferedWriter.write(String.valueOf(result));
        //bufferedWriter.newLine();
        System.out.println(result);
        //bufferedWriter.close();

        scanner.close();
    }
}

File: Interview_Kit/Search/Pairs.py
import math
import os
import random
import re
import sys

# Complete the pairs function below.
def pairs(k, arr):

    arr = sorted(arr)
    answer = 0
    # for i in range(len(arr)) :
    #     target = arr[i] + k 
    #     if target in arr : # O(n) time -> use hashset(arr[i] is unique)
    #         answer += 1
            
    # return answer
    array_set = set(arr)
    for i in range(len(arr)) :
        target = arr[i] + k
        if target in array_set : # O(1) time
            answer += 1
    return answer    
    
if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    nk = input().split()

    n = int(nk[0])

    k = int(nk[1])

    arr = list(map(int, input().rstrip().split()))

    result = pairs(k, arr)

    fptr.write(str(result) + '\n')

    fptr.close()
